underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java

```diff
@@ -0,0 +1,693 @@
+        }
+      }
+    }
+    LOG.error("Failed to copy file {} to {}, after {} retries", src, dst, retries);
+    return false;
+  }
+
+  /**
+   * Internal function to delete a key in S3.
+   *
+   * @param key the key to delete
+   * @return true if successful, false if an exception is thrown
+   */
+  private boolean deleteInternal(String key) {
+    try {
+      if (isFolder(key)) {
+        String keyAsFolder = convertToFolderName(stripPrefixIfPresent(key));
+        mClient.deleteObject(mBucketName, keyAsFolder);
+      } else {
+        mClient.deleteObject(mBucketName, stripPrefixIfPresent(key));
+      }
+    } catch (AmazonClientException e) {
+      LOG.error("Failed to delete {}", key, e);
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Gets the child name based on the parent name.
+   *
+   * @param child the key of the child
+   * @param parent the key of the parent
+   * @return the child key with the parent prefix removed, null if the parent prefix is invalid
+   */
+  private String getChildName(String child, String parent) {
+    if (child.startsWith(parent)) {
+      return child.substring(parent.length());
+    }
+    LOG.error("Attempted to get childname with an invalid parent argument. Parent: {} Child: {}",
+        parent, child);
+    return null;
+  }
+
+  /**
+   * @param key the key to get the object details of
+   * @return {@link ObjectMetadata} of the key, or null if the key does not exist
+   */
+  private ObjectMetadata getObjectDetails(String key) {
+    try {
+      if (isFolder(key)) {
+        String keyAsFolder = convertToFolderName(stripPrefixIfPresent(key));
+        return mClient.getObjectMetadata(mBucketName, keyAsFolder);
+      } else {
+        return mClient.getObjectMetadata(mBucketName, stripPrefixIfPresent(key));
+      }
+    } catch (AmazonClientException e) {
+      return null;
+    }
+  }
+
+  /**
+   * @param key the key to get the parent of
+   * @return the parent key, or null if the parent does not exist
+   */
+  private String getParentKey(String key) {
+    // Root does not have a parent.
+    if (isRoot(key)) {
+      return null;
+    }
+    int separatorIndex = key.lastIndexOf(PATH_SEPARATOR);
+    if (separatorIndex < 0) {
+      return null;
+    }
+    return key.substring(0, separatorIndex);
+  }
+
+  /**
+   * Determines if the key represents a folder. If false is returned, it is not guaranteed that the
+   * path exists.
+   *
+   * @param key the key to check
+   * @return whether the given key identifies a folder
+   */
+  private boolean isFolder(String key) {
+    // Root is always a folder
+    if (isRoot(key)) {
+      return true;
+    }
+    try {
+      String keyAsFolder = convertToFolderName(stripPrefixIfPresent(key));
+      mClient.getObjectMetadata(mBucketName, keyAsFolder);
+      // If no exception is thrown, the key exists as a folder
+      return true;
+    } catch (AmazonClientException e) {
```

Comment:
Is `AmazonClientException` the only possible exception here? how about `AmazonServiceException` ?

http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/AmazonS3Client.html#getObjectMetadata-com.amazonaws.services.s3.model.GetObjectMetadataRequest-